Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java

```diff
@@ -265,26 +273,39 @@ public synchronized  void addEmulatorCluster(EmulatorClusterDetail emulatorClust
         saveEmulatorClusters();
     }
 
-    public synchronized void addHDInsightAdditionalCluster(HDInsightAdditionalClusterDetail hdInsightClusterDetail) {
-
-        hdinsightAdditionalClusterDetails.add(hdInsightClusterDetail);
+    public synchronized void addHDInsightAdditionalCluster(@NotNull IClusterDetail hdInsightClusterDetail) {
+        assert hdInsightClusterDetail instanceof HDInsightAdditionalClusterDetail ||
+                hdInsightClusterDetail instanceof HDInsightLivyLinkClusterDetail : "trying to add a non-hdinsight-linked cluster";
+        additionalClusterDetails.add(hdInsightClusterDetail);
         ClusterMetaDataService.getInstance().addClusterToCache(hdInsightClusterDetail);
-
         saveAdditionalClusters();
     }
 
-    public synchronized  void removeEmulatorCluster(EmulatorClusterDetail emulatorClusterDetail) {
+    public synchronized void removeEmulatorCluster(EmulatorClusterDetail emulatorClusterDetail) {
         emulatorClusterDetails.remove(emulatorClusterDetail);
         ClusterMetaDataService.getInstance().removeClusterFromCache(emulatorClusterDetail);
 
         saveEmulatorClusters();
     }
 
-    public synchronized void removeHDInsightAdditionalCluster(HDInsightAdditionalClusterDetail hdInsightClusterDetail) {
-
-        hdinsightAdditionalClusterDetails.remove(hdInsightClusterDetail);
+    public synchronized void removeHDInsightAdditionalCluster(@NotNull IClusterDetail hdInsightClusterDetail) {
+        assert hdInsightClusterDetail instanceof HDInsightAdditionalClusterDetail ||
+                hdInsightClusterDetail instanceof HDInsightLivyLinkClusterDetail : "trying to remove a non-hdinsight-linked cluster";
+        additionalClusterDetails.remove(hdInsightClusterDetail);
         ClusterMetaDataService.getInstance().removeClusterFromCache(hdInsightClusterDetail);
+        saveAdditionalClusters();
+    }
 
+    public synchronized void addSqlBigDataAdditionalCluster(@NotNull IClusterDetail sqlBigDataClusterDetail) {
```

Comment:
Do we need these special methods for SqlBigDataAdditionalCluster?

Reply:
They are of no use, removed in the new commit.

---
In reply to: [237775209](https://github.com/Microsoft/azure-tools-for-java/pull/2235#discussion_r237775209) [](ancestors = 237775209)